rancher/webhook
# PodSecurityAdmissionConfigurationTemplate validation

### Issue: https://github.com/rancher/rancher/issues/39891

### Problem:
A new CRD was introduced (`PodSecurityAdmissionConfigurationTemplate`), but due to the current transition from the `v3` API to the `v1` API no validation could be written within Norman. 

### Solution: 
Add the missing validation to the web-hook inline with how other `V1` API objects perform validation.

In situations where a resource may have multiple issues with its configuration, only one error will be returned at a time. Including the description of many errors at once in the same response resulted in a less than readable result which would not display well in the UI. 

In situations where a template may be in use by either a `v1.Cluster` or `v3.Cluster` the web-hook will prevent the template from being deleted. Additionally, you cannot delete the default built-in templates `restricted` & `privileged` 

### Testing: 
A large amount of unit tests have been written, but these steps can be used for manual validation as well. 

A. Object Validation testing 
1. Create a new yaml file with the following contents 
```yaml
apiVersion: management.cattle.io/v3
kind: PodSecurityAdmissionConfigurationTemplate
metadata:
  name: testing
  namespace: testing
configuration:
  defaults:
    enforce: "privileged"
    enforce-version: "latest"
    audit: "restricted"
    audit-version: "latest"
    warn: "privileged"
    warn-version: "latest"
  exemptions:
    usernames: ["abc123"]
    runtimeClasses: ["abs"]
    namespaces: []
```
2. Create the YAML file using `kubectl`
3. Alter the YAML file in various ways to ensure validation occurs
    a. All defaults must be either 'privileged', 'baseline', or 'restricted' 
    b. All versions must be semver (`v1.x`) or `latest`
    c. All Exemptions cannot contain duplicate values 
    d. runtime classes and namespaces must be a RFC 1123 subdomain (cannot start with `-`)


B. Object deletion testing 

1. Create a new PodSecurityAdmissionConfigurationTemplate using the above YAML
2. Create an RKE1 cluster and an RKE2 cluster 
3. For both RKE1 and RKE2, Modify the cluster YAML to include the `defaultPodSecurityAdmissionConfigurationTemplateName` field directly under the spec and set its value equal to `testing` (or the name of the PSACT you created)
4. Attempt to delete the template and verify that the web-hook prevents you due to it being used by a cluster
